File: src/main/java/arriba/fix/netty/InMemoryChannelRepository.java
package arriba.fix.netty;

import java.util.Iterator;
import java.util.Map.Entry;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;

import org.jboss.netty.channel.Channel;
import org.jboss.netty.channel.ChannelFuture;
import org.jboss.netty.channel.ChannelFutureListener;

public final class InMemoryChannelRepository<ID> implements ChannelRepository<ID> {

    private final ConcurrentMap<ID, Channel> idToChannel = new ConcurrentHashMap<ID, Channel>();
    private final ChannelFutureListener removeChannelListener = new ChannelFutureListener() {
        public void operationComplete(final ChannelFuture future) throws Exception {
            InMemoryChannelRepository.this.remove(future.getChannel());
        }
    };

    public InMemoryChannelRepository() {}

    public void add(final ID id, final Channel channel) {
        final Channel previousChannel = this.idToChannel.putIfAbsent(id, channel);

        if (null == previousChannel) {
            channel.getCloseFuture().addListener(this.removeChannelListener);
        }
    }

    public void remove(final Channel channel) {
        final Iterator<Entry<ID, Channel>> iterator = this.idToChannel.entrySet().iterator();
        while (iterator.hasNext()) {
            if (channel == iterator.next().getValue()) {
                iterator.remove();

                return;
            }
        }
    }

    public void remove(final ID id) {
        this.idToChannel.remove(id);
    }

    public Channel find(final ID id) throws UnknownChannelIdException {
        final Channel channel = this.idToChannel.get(id);
        if (null == channel) {
            throw new UnknownChannelIdException("Received unknown channel ID: " + id + ".");
        }

        return channel;
    }
}

File: src/main/scala/arriba/scala/fix/netty/InMemoryChannelRepository.scala
package arriba.scala.fix.netty

import org.jboss.netty.channel.{ChannelFuture, Channel, ChannelFutureListener}
import org.jboss.netty.util.internal.ConcurrentHashMap

final class InMemoryChannelRepository[ID] extends ChannelRepository[ID] {


  def add(id: ID, channel: Channel) {
    val previousChannel: Channel = this.idToChannel.putIfAbsent(id, channel)
    if (null == previousChannel) {
      channel.getCloseFuture.addListener(this.removeChannelListener)
    }
  }

  def remove(channel: Channel) {
    val iterator = this.idToChannel.entrySet.iterator
    while (iterator.hasNext) {
      if (channel eq iterator.next.getValue) {
        iterator.remove()
        return
      }
    }
  }

  def remove(id: ID) {
    this.idToChannel.remove(id)
  }

  def find(id: ID): Channel = {
    val channel: Channel = this.idToChannel.get(id)
    if (null == channel) {
      throw new Exception("Received unknown channel ID: " + id + ".")
    }
    channel
  }

  private val idToChannel: ConcurrentHashMap[ID, Channel] = new ConcurrentHashMap[ID, Channel]
  private val removeChannelListener: ChannelFutureListener = new ChannelFutureListener {
    def operationComplete(future: ChannelFuture) {
      InMemoryChannelRepository.this.remove(future.getChannel)
    }
  }
}

